src/main/java/net/minecraftforge/event/world/JigsawPieceEvent.java

```diff
@@ -0,0 +1,57 @@
+package net.minecraftforge.event.world;
+
+import net.minecraft.util.ResourceLocation;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.gen.feature.jigsaw.JigsawPiece;
+import net.minecraftforge.eventbus.api.Event;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+
+/**
+ * Is fired from {@link net.minecraft.world.gen.feature.jigsaw.JigsawManager.Assembler#tryPlacingChildren}
+ * for each generated Jigsaw block after it stores the pool of Jigsaw pieces in can attach to.
+ * The list does NOT contain the pieces contained in {@link net.minecraft.world.gen.feature.jigsaw.JigsawPattern#fallback}
+ * to avoid infinite generation problems.
+ * <br><br>
+ * Adding a custom villager's house to plains village and zombie village can be just:
+ *
+ * {@code
+ *    String pool = event.patternLocation.getPath();
+ *    if (pool.startsWith("village/plains/houses") || pool.startsWith("village/plains/zombie/houses"))
+ *      event.pieces.add(new SingleJigsawPiece("namespace:path"));
+ * }
+ *
+ * For more detailed insight see <a href=https://mcforge.readthedocs.io/en/1.16.x/events/jigsawevent/>here.</a>
+ */
+public class JigsawPieceEvent extends Event
+{
+   public List<JigsawPiece> pieces;
+   public ResourceLocation patternLocation;
+   public BlockPos jigsawBlockPosition;
+   public Map<String, Object> jigsawAssemblerSettings;
+   public Random random;
+   public final int currentDepth;
+
+   /**
+    * @param pieces                  The list of pieces that have a probability of being placed. Any addition to big pools should be shuffled for better results.
+    *                                The list does NOT contain the fallback pool of the current pool. See the appropriate pool for the structure to know what the fallback.
+    *                                ({@link net.minecraft.world.gen.feature.structure.PlainsVillagePools} for Plains Village)
+    * @param patternLocation         The resource location of the {@link net.minecraft.world.gen.feature.jigsaw.JigsawPattern}.
+    * @param jigsawBlockPosition     The block position of jigsaw block requesting the pieces.
+    * @param jigsawAssemblerSettings The map of generate-time settings for specific structure. It's unused for vanilla structures but can be used
+    *                                in modded ones to add specific settings or more or less randomness.
+    * @param random                  The access to Random instance of specific structure. Can be used to shuffle custom pieces after they are added.
+    * @param currentDepth            The depth of the piece that is going to be placed. It will be 0 for the jigsaw blocks contained in the starting piece.
+    */
+   public JigsawPieceEvent(List<JigsawPiece> pieces, ResourceLocation patternLocation, BlockPos jigsawBlockPosition, Map<String, Object> jigsawAssemblerSettings, Random random, int currentDepth)
```

Comment:
should this be cancelable?

If you clear the list to generate a diamond house and sets `true` into the `jigsawAssemblerSettings` and another mod thinks the same, only one is generated, but both thought they have generated the jigsawpiece

Reply:
Maybe. I was more thinking about using this with modded structures though so you don't have to worry about other mod doing something with your structure. But for vanilla structures, something should probably be done.

Also, if event is cancelled, does this mean all future events handlers will not be called, even for other structure? If so, something should probably be changed to allow cancelling either only events for this specific structure instance or only all events for this structure type.

Reply:
that just means that this specific event will not be called for other event subscriber and the next JigsawEvent that is fired will be again called for every event subscriber. (if that makes it clear)